clients/pkg/promtail/client/client.go

```diff
@@ -451,11 +458,30 @@ func (c *client) processEntry(e api.Entry) (api.Entry, string) {
 	return e, tenantID
 }
 
-func (c *client) UnregisterLatencyMetric(labels prometheus.Labels) {
-	labels[HostLabel] = c.cfg.URL.Host
-	c.metrics.streamLag.Delete(labels)
+func (c *client) UnregisterLatencyMetric(labels prometheus.Labels) bool {
+	key := c.streamLagMetricsMapKey(labels)
+	lset, ok := c.streamLagMetricsMap[key]
+	if !ok {
+		level.Debug(c.logger).Log("msg", "no stream lag metric found", "filename", labels["filename"], "client", c.name, "host", c.cfg.URL.Host)
+		return false
+	}
+	// If we found an entry in the map it's always safe to delete it, this is the
+	// only place we try to look for that entry. This guarantees we don't endlessly
+	// grow the map entries + if the entire labelset is seen again we would just
+	// write this entry back into the map the next time we send a batch for the stream.
+	delete(c.streamLagMetricsMap, key)
+
+	if !c.metrics.streamLag.Delete(lset) {
+		level.Debug(c.logger).Log("msg", "no stream lag metric deleted, mismatch between stored metrics map and client library registered metrics", "filename", labels["filename"], "client", c.name, "host", c.cfg.URL.Host)
+		return false
+	}
+	return true
 }
 
 func (c *client) Name() string {
 	return c.name
 }
+
+func (c *client) streamLagMetricsMapKey(labels prometheus.Labels) string {
```

Comment:
Does this happen before relabeling/etc? If a config drops the `filename` label, is this ok?

Reply:
Good catch. This happens after relabelling.

By the time we receive an entry on the clients channel the streams labelset has already had relabelling rules applied. So then we can't guarantee that we have the filename (or other unique target identifying label).

Without my recent PRs we expose the same metric twice, which results in Prometheus being unable to scrape `/metrics`. With my recent PRs and someone using relabelling or the `labeldrop` pipeline stage they would likely end up with one stream lag metric per scrape job. That doesn't change with this PR. We would simply track all stream lag for files with `filename` dropped as if they were just one metric for the client.

This PR would allow us to at least not continue reporting the stream lag metric for a file that goes away in **most** cases. We still need additional changes/refactoring to handle all cases for this metric properly. Ed and I are going to discuss further on Monday.